Sources/NIO/PendingWritesManager.swift

```diff
@@ -447,13 +447,92 @@ internal protocol PendingWritesManager {
     var isFlushPending: Bool { get }
     var writeSpinCount: UInt { get }
     var currentBestWriteMechanism: WriteMechanism { get }
-    var channelWritabilityFlag: Atomic<Bool> { get }
+    var unwritableFlags: Atomic<UInt64> { get }
 }
 
 extension PendingWritesManager {
     // This is called from `Channel` API so must be thread-safe.
     var isWritable: Bool {
-        return self.channelWritabilityFlag.load()
+        return self.unwritableFlags.load() == 0
+    }
+
+    internal func setWritable() -> Bool {
+        // Check if flag already cleared
+        if unwritableFlags.load() & 1 == 0 {
+            return false
+        }
+        var oldValue: UInt64
+        var newValue: UInt64
+        repeat {
+            oldValue = unwritableFlags.load()
+            newValue = oldValue & UInt64(bitPattern: -2);
+        }
+        while !unwritableFlags.compareAndExchange(expected: oldValue, desired: newValue)
+        return true
+    }
+
+    internal func setUnwritable() -> Bool {
+        // Check if flag already cleared
+        if unwritableFlags.load() & 1 == 1 {
+            return false
+        }
+        var oldValue: UInt64
+        var newValue: UInt64
+        repeat {
+            oldValue = unwritableFlags.load()
+            newValue = oldValue | 1
+        }
+        while !unwritableFlags.compareAndExchange(expected: oldValue, desired: newValue)
+        return true
+    }
+
+    internal func getUserDefinedWritability(index: Int) -> Bool {
+        let mask = writabilityMask(index)
+        return (unwritableFlags.load() & mask) == mask
+    }
+
+    internal func setUserDefinedWritability(index: Int, writable: Bool) -> Bool {
+        if (writable) {
+            return setUserDefinedWritability(index: index)
+        }
+        else {
+            return clearUserDefinedWritability(index: index)
+        }
+    }
+
+    internal func setUserDefinedWritability(index: Int) -> Bool {
+        let mask = ~writabilityMask(index)
+        var oldValue: UInt64
+        var newValue: UInt64
+        repeat {
+            oldValue = unwritableFlags.load()
+            newValue = oldValue & mask
+        }
+        while !unwritableFlags.compareAndExchange(expected: oldValue, desired: newValue)
+
+        return oldValue != 0 && newValue == 0
+    }
+
+    internal func clearUserDefinedWritability(index: Int) -> Bool {
+        let mask = writabilityMask(index)
+        var oldValue: UInt64
+        var newValue: UInt64
+        repeat {
+            oldValue = unwritableFlags.load()
+            newValue = oldValue | mask
+        }
+        while !unwritableFlags.compareAndExchange(expected: oldValue, desired: newValue)
+
+        return oldValue == 0 && newValue != 0
+    }
```

Comment:
So we've duplicated the basic body of this method 4 times, differing only on the bitwise ops and the mask.

We can and should begin by dropping this to two implementations that both accept a writability mask, one for setting and one for unsetting the mask. Each of these 4 functions can call that base implementation.

I don't think the advantage of going to one implementation helps us, though, as we'd need to pass closures in, which will be a bit slower than we really want at the moment.